From 87b9f374053a786e54f44c2a9ea93e582ffad606 Mon Sep 17 00:00:00 2001
From: Sailaja <sbandarupalli@habana.ai>
Date: Thu, 01 Oct 2020 14:58:14 +0300
Subject: [PATCH] [SW-21652] goya2: init sram scrambling

starting Goya2, it is decided that firmware takes care of
enabling DCON SRAM scrambling instead of LKD

Change-Id: I5914cece1f21f2a17ecd2e615a1984f6e1889766
Signed-off-by: Sailaja <sbandarupalli@habana.ai>
---

diff --git a/goya2/asic.c b/goya2/asic.c
index 972527c..fe1f864 100644
--- a/goya2/asic.c
+++ b/goya2/asic.c
@@ -7,6 +7,14 @@
 
 #include <psoc_global_conf_regs.h>
 #include <psoc_gpio0_regs.h>
+#include <dcon0_hbw_rtr_if0_rtr_h3_regs.h>
+#include <dcon0_hbw_rtr_if1_rtr_h3_regs.h>
+#include <dcon1_hbw_rtr_if0_rtr_h3_regs.h>
+#include <dcon1_hbw_rtr_if1_rtr_h3_regs.h>
+#include <dcon2_hbw_rtr_if0_rtr_h3_regs.h>
+#include <dcon2_hbw_rtr_if1_rtr_h3_regs.h>
+#include <dcon3_hbw_rtr_if0_rtr_h3_regs.h>
+#include <dcon3_hbw_rtr_if1_rtr_h3_regs.h>
 
 #include <preboot.h>
 #include <pll.h>
@@ -62,3 +70,16 @@
 	val &= ~BIT(2);
 	regmap_writel(val, mmPSOC_GPIO0_DATAOUT);
 }
+
+/* initialize sram scrambling before u-boot is downloaded */
+void sram_scrambling_init(void)
+{
+	regmap_writel(1, mmDCON0_HBW_RTR_IF0_RTR_H3_SCRAM_SRAM_EN);
+	regmap_writel(1, mmDCON0_HBW_RTR_IF1_RTR_H3_SCRAM_SRAM_EN);
+	regmap_writel(1, mmDCON1_HBW_RTR_IF0_RTR_H3_SCRAM_SRAM_EN);
+	regmap_writel(1, mmDCON1_HBW_RTR_IF1_RTR_H3_SCRAM_SRAM_EN);
+	regmap_writel(1, mmDCON2_HBW_RTR_IF0_RTR_H3_SCRAM_SRAM_EN);
+	regmap_writel(1, mmDCON2_HBW_RTR_IF1_RTR_H3_SCRAM_SRAM_EN);
+	regmap_writel(1, mmDCON3_HBW_RTR_IF0_RTR_H3_SCRAM_SRAM_EN);
+	regmap_writel(1, mmDCON3_HBW_RTR_IF1_RTR_H3_SCRAM_SRAM_EN);
+}
diff --git a/goya2_arc_gnu/asic.c b/goya2_arc_gnu/asic.c
index 2164113..07fe02c 100644
--- a/goya2_arc_gnu/asic.c
+++ b/goya2_arc_gnu/asic.c
@@ -7,6 +7,14 @@
 
 #include <psoc_global_conf_regs.h>
 #include <psoc_gpio0_regs.h>
+#include <dcon0_hbw_rtr_if0_rtr_h3_regs.h>
+#include <dcon0_hbw_rtr_if1_rtr_h3_regs.h>
+#include <dcon1_hbw_rtr_if0_rtr_h3_regs.h>
+#include <dcon1_hbw_rtr_if1_rtr_h3_regs.h>
+#include <dcon2_hbw_rtr_if0_rtr_h3_regs.h>
+#include <dcon2_hbw_rtr_if1_rtr_h3_regs.h>
+#include <dcon3_hbw_rtr_if0_rtr_h3_regs.h>
+#include <dcon3_hbw_rtr_if1_rtr_h3_regs.h>
 
 #include <preboot.h>
 #include <pll.h>
@@ -48,3 +56,16 @@
 	val &= ~BIT(2);
 	regmap_writel(val, mmPSOC_GPIO0_DATAOUT);
 }
+
+/* initialize sram scrambling before u-boot is downloaded */
+void sram_scrambling_init(void)
+{
+	regmap_writel(1, mmDCON0_HBW_RTR_IF0_RTR_H3_SCRAM_SRAM_EN);
+	regmap_writel(1, mmDCON0_HBW_RTR_IF1_RTR_H3_SCRAM_SRAM_EN);
+	regmap_writel(1, mmDCON1_HBW_RTR_IF0_RTR_H3_SCRAM_SRAM_EN);
+	regmap_writel(1, mmDCON1_HBW_RTR_IF1_RTR_H3_SCRAM_SRAM_EN);
+	regmap_writel(1, mmDCON2_HBW_RTR_IF0_RTR_H3_SCRAM_SRAM_EN);
+	regmap_writel(1, mmDCON2_HBW_RTR_IF1_RTR_H3_SCRAM_SRAM_EN);
+	regmap_writel(1, mmDCON3_HBW_RTR_IF0_RTR_H3_SCRAM_SRAM_EN);
+	regmap_writel(1, mmDCON3_HBW_RTR_IF1_RTR_H3_SCRAM_SRAM_EN);
+}
diff --git a/preboot/preboot.c b/preboot/preboot.c
index 0e8221c..5882bfa 100644
--- a/preboot/preboot.c
+++ b/preboot/preboot.c
@@ -186,6 +186,7 @@
 
 __weak void preboot_security_init(void) {}
 __weak void preboot_led_on(void) {}
+__weak void sram_scrambling_init(void) {}
 
 void preboot_main(void)
 {
@@ -208,5 +209,6 @@
 	preboot_version();
 	fuse_version();
 	preboot_led_on();
+	sram_scrambling_init();
 	module_handle_cmds(cmd_ops, ARRAY_SIZE(cmd_ops));
 }
